core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -221,28 +222,172 @@ async def process_internal_input(self, txi: TxInput) -> None:
     async def process_external_input(self, txi: TxInput) -> None:
         self.approver.add_external_input(txi)
 
-    async def approve_output(self, txo: TxOutput, script_pubkey: bytes) -> None:
-        if self.output_is_change(txo):
-            # output is change and does not need approval
+    async def process_original_input(self, txi: TxInput) -> None:
+        assert txi.orig_hash is not None
+        assert txi.orig_index is not None
+
+        for orig in self.orig_txs:
+            if orig.orig_hash == txi.orig_hash:
+                break
+        else:
+            orig_meta = await helpers.request_tx_meta(
+                self.tx_req, self.coin, txi.orig_hash
+            )
+            orig = OriginalTxInfo(self, orig_meta, txi.orig_hash)
+            self.orig_txs.append(orig)
+
+        if txi.orig_index >= orig.tx.inputs_count:
+            raise wire.ProcessError("Not enough inputs in original transaction.")
+
+        if orig.index != txi.orig_index:
+            raise wire.ProcessError(
+                "Rearranging or removal of original inputs is not supported."
+            )
+
+        orig_txi = await helpers.request_tx_input(
+            self.tx_req, txi.orig_index, self.coin, txi.orig_hash
+        )
+
+        if (
+            orig_txi.prev_hash != txi.prev_hash
+            or orig_txi.prev_index != txi.prev_index
+            or orig_txi.amount != txi.amount
+        ):
+            raise wire.ProcessError("Original input does not match current input.")
+
+        orig.add_input(orig_txi)
+        orig.index += 1
+
+    async def fetch_removed_original_outputs(
+        self, orig: OriginalTxInfo, orig_hash: bytes, last_index: int
+    ) -> None:
+        while orig.index < last_index:
+            txo = await helpers.request_tx_output(
+                self.tx_req, orig.index, self.coin, orig_hash
+            )
+            orig.add_output(txo, self.output_derive_script(txo))
+
+            if orig.output_is_change(txo):
+                # Removal of change-outputs is allowed.
+                self.approver.add_orig_change_output(txo)
+            else:
+                # Removal of external outputs requires prompting the user. Not implemented.
+                raise wire.ProcessError(
+                    "Removal of original external outputs is not supported."
+                )
+                self.approver.add_orig_external_output(txo)
```

Comment:
> actually this is an important piece of logic that conceptually belongs to the Approver; namely that removing or reducing external outputs is not supported.

I agree, that it would be nice to have all of this logic in one place, but I am not sure to what extent that is practical. In this particular case I considered that we already handle "Removal of original inputs is not supported." in bitcoin.py and it doesn't seem to make a whole lot of sense to move it to the approver. Sure, we can add a function like `add_removed_input(txi)` or something like that to the approver, which will just raise an exception, but then there is "Rearranging or removal of original inputs is not supported." somewhere else and I am not sure whether involving the approver makes sense there. So basically I just decided to handle all removals in bitcoin.py. Maybe there is an elegant solution that I am not seeing right now.

> I'm not sure about a complete solution, but how about changing signature of `Approver.add_change_output` and `Approver.add_external_output` so that the "current" output is also optional. Then output removal is indicated by passing `txo=None, orig_txo=txo`, and the approver can do whatever it needs.

At some point I was actually trying to do something like that, but it fell apart when I realized that an output which is a change-output in the original might not be classified as a change output in the replacement. Like if the original transaction is all singlesig and you add an internal multisig input to the replacement transaction.

So I think the only alternative to the current setup that I thought of was having just one `add_output()` function:
```
add_output(
    self,
    script_pubkey: bytes,
    txo: Optional[TxOutput],
    is_change: bool = False,
    orig_txo: Optional[TxOutput] = None,
    orig_is_change: bool = False,
)
```
It seems like a viable option but I haven't considered all the consequences.

Reply:
OK, I agree that the current solution is more practical.
Please remove the approver call though :)

Reply:
> Please remove the approver call though :)

I forgot to mention, I removed that in https://github.com/trezor/trezor-firmware/pull/1292/commits/fdd2e91dcdf0211750cf276dc9515a081390763d